File: mpif/src/main/java/com/ivir/mpif/simdata/CasualtyState.java
/*
 * Copyright 2026 IVIR Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ivir.mpif.simdata;

public class CasualtyState extends ConcurrentSimData<CasualtyState.Attributes>{
    public enum Attributes{
        ID,
        LOCAL_BOOL,
        INSTANCE_NAME_STR,
        PATIENT_ID_STR,
        FACILITY_ID_STR,
        EVACUATION_PRIORITY_ENUM,
        TRIAGE_CLASSIFICATION_ENUM,
    }

    public CasualtyState(){
        super(Attributes.class, Attributes.ID, true);
    }

    public String getId(){
        return this.getIndex();
    }

    public Boolean isLocal(){
        return this.getValue(Attributes.LOCAL_BOOL, Boolean.class);
    }

    public CasualtyState setLocal(Boolean isLocal){
        this.setValue(Attributes.LOCAL_BOOL, isLocal);
        return this;
    }

    public String getFacilityId(){
        return this.getValue(Attributes.FACILITY_ID_STR, String.class);
    }

    public CasualtyState setFacilityId(String facilityId){
        this.setValue(Attributes.FACILITY_ID_STR, facilityId);
        return this;
    }

    public String getInstanceName(){
        return this.getValue(Attributes.INSTANCE_NAME_STR, String.class);
    }

    public CasualtyState setInstanceName(String name){
        this.setValue(Attributes.INSTANCE_NAME_STR, name);
        return this;
    }

    public String getPatientId(){
        return this.getValue(Attributes.PATIENT_ID_STR, String.class);
    }

    public CasualtyState setPatientId(String patientId){
        this.setValue(Attributes.PATIENT_ID_STR, patientId);
        return this;
    }

    public EvacuationPriority getEvacuationPriority(){
        return this.getValue(Attributes.EVACUATION_PRIORITY_ENUM, EvacuationPriority.class);
    }

    public CasualtyState setEvacuationPriority(EvacuationPriority evacuationPriority){
        this.setValue(Attributes.EVACUATION_PRIORITY_ENUM, evacuationPriority);
        return this;
    }

    public TriageClassification getTriageClassification(){
        return this.getValue(Attributes.TRIAGE_CLASSIFICATION_ENUM, TriageClassification.class);
    }

    public CasualtyState setTriageClassification(TriageClassification triageClassification){
        this.setValue(Attributes.TRIAGE_CLASSIFICATION_ENUM, triageClassification);
        return this;
    }

}

File: flutter_ui/lib/data/model/casualty_state/casualty_state.dart
/*
 * Copyright 2026 IVIR Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'package:equatable/equatable.dart';
import 'package:flutter_ui/data/storage/base_data_store.dart';

class CasualtyState extends Equatable {
  final String id;
  final String patientId;
  final String evacuationPriority; // backend values (e.g., ROUTINE)
  final String triageClassification; // backend values (e.g., DELAYED)
  final String facilityId; // String; same on front and backend

  const CasualtyState({
    required this.id,
    required this.patientId,
    required this.evacuationPriority,
    required this.triageClassification,
    required this.facilityId,
  });

  CasualtyState copyWith({
    String? id,
    String? patientId,
    String? evacuationPriority,
    String? triageClassification,
    String? facilityId,
  }) {
    return CasualtyState(
      id: id ?? this.id,
      patientId: patientId ?? this.patientId,
      evacuationPriority: evacuationPriority ?? this.evacuationPriority,
      triageClassification: triageClassification ?? this.triageClassification,
      facilityId: facilityId ?? this.facilityId,
    );
  }

  @override
  List<Object?> get props =>
      [id, patientId, evacuationPriority, triageClassification, facilityId];
}

class CasualtyStateWireStore extends BaseDataStore {
  final void Function(CasualtyState cs) onUpsert;

  CasualtyStateWireStore({
    required this.onUpsert,
  });

  @override
  bool acceptsJson() => true;
  @override
  bool acceptsRawString() => false;
  @override
  bool acceptsBool() => false;
  @override
  bool acceptsInt() => false;
  @override
  bool acceptsDouble() => false;

  @override
  void parseAndStoreJsonByDataType(String dataType, dynamic payload) {
    if (dataType != 'CasualtyState') return;
    _ingest(payload);
  }

  void _ingest(dynamic payload) {
    if (payload is List) {
      for (final item in payload) _ingest(item);
      return;
    }
    if (payload is! Map<String, dynamic>) return;

    final patientId = (payload['patientId'] ?? '').toString();
    if (patientId.isEmpty) return;

    // Optional extras we currently don't persist (available if you want them later)
    final bool isLocal = payload['local'] == true; // true
    final String? instanceName = payload['instanceName'] as String?; // null

    // Map only the fields your model has today
    final cs = CasualtyState(
      id: (payload['id'] ?? 'ws-$patientId').toString(), // "0"
      patientId: patientId,
      evacuationPriority:
          (payload['evacuationPriority'] ?? '').toString(), // "ROUTINE"
      triageClassification:
          (payload['triageClassification'] ?? '').toString(), // "MINIMAL"
      facilityId:
          (payload['facilityId'] ?? '').toString(), // "" → Unassigned in UI
    );

    // You can log/use isLocal/instanceName here if you ever need to.
    onUpsert(cs);
  }

  // Unused for this store
  @override
  void storeRawStringByDataType(String d, String i, String r) {}
  @override
  void storeBoolByDataType(String d, String i, bool v) {}
  @override
  void storeIntByDataType(String d, String i, int v) {}
  @override
  void storeDoubleByDataType(String d, String i, double v) {}
}
